### DISCUSSION TITLE
BIRT-runtime (OSGI) manual replacement of batik to convert SVG to JPEG

### DISCUSSION THREAD

**Discussion by speckyspooky** [eclipse-birt/birt#805] [Q&A] [answered]
Hello together,

I have an question to the working of the BIRT-runtime "ReportEngine with OSGI".

Currently I have a very special report which create an SVG-string-buffer which will be converted to an JPEG at the end of processing.
My verified/tested versions are BIRT 4.6 (also tested: 4.7, 4.8 with Java 1.8) on Tomcat with BIRT-Viewer.
On default it isn't possible to convert the SVG to JPEG because issues with the old batik-libraries. 
So I added the "batik-all"-version (tested batik 1.11, 1.12, 1.14) to the BIRT-lib-folder and removed the single libraries "org.apache.batik.*".
(I know this kind of replacement isn't the best way.)

The result is, the report can be created perfectly and the SVG will be converted correct to an JPEG-image.

**My problem begins with BIRT 4.6 OSGI:** 
I have a server-portal which use the "BIRT 4.6 OSGI"-version based on type "ReportEngine" (not used with BIRT-Viewer).
I replaced the old-batik-libraries in the same way.
But when I try to start the report creation the ReportEngine gives me an fatal exception with the message "ReportEngine cannot be initialized" and crashs total without more information.

**My question is:**
Is there a way to configure the ReportEngine to use the new batik-version without my (bad) replacement of libs?
What could be the reason that the BIRT-Viewer works perfect in all combinations 4.6-4.8 with 1.11-1.14 but not the combination of ReportEngine with OSGI?

I attached a screen of the converting functio of the report:
![BIRT-4 6-SVG-to-JPEG-converting](https://user-images.githubusercontent.com/41593722/152036985-c0ee616b-4959-4a18-b229-a9632f157987.png)

It would be great if someone could help me with some hints on it.

Best regards
Thomas

PS.: BIRT 4.9 has the perfect modern version of batik - I hope the release is coming soon.

**Comment by wimjongman** [comment, marked as answer]
In OSGi you cannot simply replace jars and hope for the best. The resolution of the classpath is what OSGi is all about.

You also have to replace the bundles in ReportEngine\platform\configuration\org.eclipse.equinox.simpleconfigurator\bundles.info

**Reply by speckyspooky** [reply]
I verified the 4.6-OSGi and there isn't such folder structure by default incl. the bundle.info.
I compared it with 4.8.OSGi and there is the structure and the bundles.info given with the full list of used plugins.

Do you know whether 4.6-OSGi use the "simpleconfigurator" because I didn't found such library.
And If I create the "bundle.info" manual must be listed all plugins there or only my changes for batik?

**Reply by wimjongman** [reply]
Also take a look at the various plugins folders in 4.6. There is one in the WebViewerExample/WEB-INF and one in the ReportEngine. 4.6 does not use the bundles.info but can detect the bundles automatically from the plugins folder.

**Comment by hvbtup** [comment]
In case you create PDF output with BIRT, you also use an image item to render the SVG as PDF vector graphics.
I am using this technique in some reports as a replacement for the RotatedTextItem, which has some issues (often too tiny or low-res, depending on the report's DPI setting IIRC).
